src/main/java/de/tum/in/www1/artemis/service/quiz/QuizResultService.java

```diff
@@ -0,0 +1,190 @@
+    }
+
+    /**
+     * // @formatter:off
+     * Evaluate the given quiz exercise by performing the following actions for each participation:
+     * 1. Get the submission for each participation (there should be only one as in exam mode, the submission gets created upfront and will be updated)
+     * - If no submission is found, print a warning and continue as we cannot evaluate that submission
+     * - If more than one submission is found, select one of them
+     * 2. mark submission and participation as evaluated
+     * 3. Create a new result for the selected submission and calculate scores
+     * 4. Save the updated submission & participation and the newly created result
+     * <p>
+     * After processing all participations, the created results will be returned for further processing
+     * Note: We ignore test run participations
+     * // @formatter:on
+     *
+     * @param quizExercise the id of the QuizExercise that should be evaluated
+     * @return the newly generated results
+     */
+    private Set<Result> evaluateSubmissions(@NotNull QuizExercise quizExercise) {
+        Set<Result> createdResults = new HashSet<>();
+        List<StudentParticipation> studentParticipations = studentParticipationRepository.findAllWithEagerLegalSubmissionsAndEagerResultsByExerciseId(quizExercise.getId());
+        submittedAnswerRepository.loadQuizSubmissionsSubmittedAnswers(studentParticipations);
+
+        for (var participation : studentParticipations) {
+            if (!participation.isTestRun()) {
+                try {
+                    // reconnect so that the quiz questions are available later on (otherwise there will be a org.hibernate.LazyInitializationException)
+                    participation.setExercise(quizExercise);
+                    Set<Submission> submissions = participation.getSubmissions();
+                    QuizSubmission quizSubmission;
+                    if (submissions.isEmpty()) {
+                        log.warn("Found no submissions for participation {} (Participant {}) in quiz {}", participation.getId(), participation.getParticipant().getName(),
+                                quizExercise.getId());
+                        continue;
+                    }
+                    else if (submissions.size() > 1) {
+                        log.warn("Found multiple ({}) submissions for participation {} (Participant {}) in quiz {}, taking the one with highest id", submissions.size(),
+                                participation.getId(), participation.getParticipant().getName(), quizExercise.getId());
+                        List<Submission> submissionsList = new ArrayList<>(submissions);
+
+                        // Load submission with highest id
+                        submissionsList.sort(Comparator.comparing(Submission::getId).reversed());
+                        quizSubmission = (QuizSubmission) submissionsList.getFirst();
```

Comment:
```suggestion
                        quizSubmission = (QuizSubmission) submissions.stream().max(Comparator.comparing(Submission::getId)).get();
```
A matter of taste, but I think this is way shorter and more understandable